### PR TITLE
Extend the parser to accept effect annotations.

### PR DESCRIPTION
**Pull request by eldesh**
#169
Extend the parser to accept effect annotations.

Here is an example of what it will be able to accept.

```ante
foo (f: Unit -> a can GiveInt) : a = ...

foo () : Unit can Throw FileError, Throw ParseError, Throw BarError = ...

foo () can Throw FileError, Throw ParseError, Throw BarError = ...
```

No semantic analysis is done, just parsing.

### CODE REVIEW

File: examples/parsing/effects.an [hunk 1]
```diff
@@ -24,6 +24,23 @@ handle ()
 | return _ -> "foo"
 | log _ -> "impossible"
```
Comment (eldesh) [inline examples/parsing/effects.an:26]: ported from language tour.

File: src/parser/ast.rs [hunk 2]
```diff
@@ -205,13 +206,19 @@ pub enum Type<'a> {
     Boolean(Location<'a>),
     Unit(Location<'a>),
     Reference(Location<'a>),
-    Function(Vec<Type<'a>>, Box<Type<'a>>, /*varargs:*/ bool, /*closure*/ bool, Location<'a>),
+    Function(Vec<Type<'a>>, Box<Type<'a>>, Vec<Effect<'a>>, /*varargs:*/ bool, /*closure*/ bool, Location<'a>),
```
Comment (eldesh) [inline src/parser/ast.rs:209]: Is non-ordered container preferable?

Comment (jfecher) [reply src/parser/ast.rs:209]: Likely, but this is fine as a first pass. The order of effect annotations doesn't affect anything, but the order of effect handlers in the code can affect the output.

File: src/parser/mod.rs [hunk 3]
```diff
@@ -116,6 +117,8 @@ parser!(varargs location -> 'b () =
     ()
 );
 
+// Parses the return type of function.
+// This may involve effects.
```
Comment (jfecher) [inline src/parser/mod.rs:121]:
```suggestion
```
This comment seems outdated seeing as the `function_return_type` parser is separate from the `effect_set` parser in `function_definition`

File: src/parser/pretty_printer.rs [hunk 4]
```diff
@@ -53,6 +53,9 @@ impl<'a> Display for ast::Lambda<'a> {
         if let Some(typ) = &self.return_type {
             write!(f, " : {}", typ)?;
         }
+        if self.effects.len() > 0 {
+          write!(f, " can {}", join_with(&self.effects, ","))?;
```
Comment (jfecher) [inline src/parser/pretty_printer.rs:57]:
```suggestion
          write!(f, " can {}", join_with(&self.effects, ", "))?;
```

**Review by jfecher — APPROVED** [review]
Looks good, thanks!

I think one change I'll add in a separate commit is to have the effect_set parser return an `Option<Vec<...>>` instead of just a `Vec<...>` so that we can differentiate between "user did not specify an effect set" and "user specified there should be no effects." Thanks for the help on this issue!

**Pull request merged**